mne/decoding/gsoc.py

```diff
@@ -0,0 +1,280 @@
+"""Contains refactored transformers compatible with scikit-learn"""
+
+import numpy as np
+import warnings
+from scipy import linalg
+
+from .. import Covariance
+from ..cov import _regularized_covariance
+from ..preprocessing import Xdawn
+
+
+class XdawnTransformer(Xdawn):
+
+    """Implementation of the Xdawn Algorithm.
+
+    Xdawn is a spatial filtering method designed to improve the signal
+    to signal + noise ratio (SSNR) of the ERP responses. Xdawn was originally
+    designed for P300 evoked potential by enhancing the target response with
+    respect to the non-target response. This implementation is a light version
+    of the original version and follows scikit-learn API strictly.
+
+    .. note:: This does not handle overlapping events. Use original
+              preprocessing.xdawn instead.
+
+    Parameters
+    ----------
+    n_components : int (default 2)
+        The number of components to decompose M/EEG signals.
+    n_chan : int
+         Integer indicating the number of channels.
+    signal_cov : None | Covariance | ndarray, shape (n_channels, n_channels)
+        (default None). The signal covariance used for whitening of the data.
+        if None, the covariance is estimated from the epochs signal.
+    reg : float | str | None (default None)
+        if not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
+        or Oracle Approximating Shrinkage ('oas').
+
+    Attributes
+    ----------
+    filters_ : dict of ndarray
+        If fit, the Xdawn components used to decompose the data for each event
+        type, else empty.
+    patterns_ : dict of ndarray
+        If fit, the Xdawn patterns used to restore M/EEG signals for each event
+        type, else empty.
+
+
+    See Also
+    --------
+    mne.preprocessing.Xdawn
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, n_chan=None, signal_cov=None,
```

Comment:
This whole class is copied and pasted from the original Xdawn no? If so, can you try to make a private class that both Xdawn and XdawnTransformer access to avoid code redundancy?